src/diffusers/models/unet_2d_blocks.py

```diff
@@ -2198,3 +2440,326 @@ def forward(
+
+        for resnet, attn in zip(self.resnets, self.attentions):
+            if self.training and self.gradient_checkpointing:
+
+                def create_custom_forward(module, return_dict=None):
+                    def custom_forward(*inputs):
+                        if return_dict is not None:
+                            return module(*inputs, return_dict=return_dict)
+                        else:
+                            return module(*inputs)
+
+                    return custom_forward
+
+                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                hidden_states = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(attn, return_dict=False),
+                    hidden_states,
+                    encoder_hidden_states,
+                    attention_mask,
+                    cross_attention_kwargs,
+                )[0]
+            else:
+                hidden_states = resnet(hidden_states, temb)
+                hidden_states = attn(
+                    hidden_states,
+                    encoder_hidden_states=encoder_hidden_states,
+                    emb=temb,
+                    attention_mask=attention_mask,
+                    cross_attention_kwargs=cross_attention_kwargs,
+                )
+
+        if self.upsamplers is not None:
+            for upsampler in self.upsamplers:
+                hidden_states = upsampler(hidden_states)
+
+        return hidden_states
+
+
+# can potentially later be renamed to `No-feed-forward` attention
+class KAttentionBlock(nn.Module):
+    r"""
+    A basic Transformer block.
+
+    Parameters:
+        dim (`int`): The number of channels in the input and output.
+        num_attention_heads (`int`): The number of heads to use for multi-head attention.
+        attention_head_dim (`int`): The number of channels in each head.
+        dropout (`float`, *optional*, defaults to 0.0): The dropout probability to use.
+        cross_attention_dim (`int`, *optional*): The size of the encoder_hidden_states vector for cross attention.
+        activation_fn (`str`, *optional*, defaults to `"geglu"`): Activation function to be used in feed-forward.
+        num_embeds_ada_norm (:
+            obj: `int`, *optional*): The number of diffusion steps used during training. See `Transformer2DModel`.
+        attention_bias (:
+            obj: `bool`, *optional*, defaults to `False`): Configure if the attentions should contain a bias parameter.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_attention_heads: int,
+        attention_head_dim: int,
+        dropout: float = 0.0,
+        cross_attention_dim: Optional[int] = None,
+        attention_bias: bool = False,
+        upcast_attention: bool = False,
+        temb_channels: int = 768,  # for ada_group_norm
+        attn1_type: str = "cross",  # cross, self
```

Comment:
Can we try to get rid of those? 

If I understand correctly there are only two different "type" combinations here. 
Either we have:
- just a single cross attention layer + AdaGroupNorm
- a self-attention layer + AdaGroupNorm + a cross-attention layer + AdaGroupNorm

=> So could we maybe remove both `attn1_type` and `attn2_type` and instead have one `add_self_attention=True/False` parameter that defines whether `self.attn1` is None or not. `self.attn2` can then **always** be the cross attention layer.
This has a couple of advantages:
- Easier to read, one knows directly that there is always a cross attention layer.
- We make it clear that `attn1` is always a self-attention layer which is more similar to the original attention is all you need architecture
- This attention class is very specific for `KUp` and `KDown` so IMO here it's good to use a easy to understand boolean instead of a complex mixture of strings. 


Maybe to explain a bit better. For things like normalization layers, activation functions, unet blocks it's better to use a **single** string configuration variable, e.g. 
- "norm_type": {"layer", "batch", "group", "ada_group", ...}
- "act_fn": {relu, gelu, selu, ...}
- "block_type": {"ResNetConv", "ResNet", "ResNetAttention", "ResNetCrossAttention"}

They all already have a lot of different variations that more or less all do the same task, API and it can be expected that they will grow in variety:
- accept hidden states and return normalized hidden states
- accept hidden states and return "non-linear" activation of hidden states
- accept hidden image tensor and return downsampled hidden image tensor (unet blocks also with an * here because they don't always accept the exact same inputs)
- We will add more unet blocks, we will add more activation functions,...

And with the exception of the unet blocks code stays very readable when using `"type"` configurations. E.g.
```py
self.act_fn = "..."
hidden_states = self.act_fn(hidden_states)
```
is easy to understand without knowing what the act fn is.

In contrast, "attn1_type" doesn't tick all of the above boxes:
- there is only cross attention and self attention
- The API and the whole purpose of the layer is not the same for self attention and cross attention. Cross attention only works with two inputs, self-attention only works with one input
- There is very little reason to believe that we will have more variation than just self and cross anytime soon.
- IMO it's not that easy to directly know what's going on with:
```py
self.attn = 
hidden_states = self.attn(hidden_states, ...)
````

if we don't know what `self.attn = ...` is.

Also note that the above is not a 100% sound because one can easily argue that the design of "KCrossAttnDownBlock2D" is not really readable and it's hard to understand the logic when reading `up_blocks=["KCrossAttnDownBlock2D", "KCrossAttnDownBlock2D", "KCrossAttnDownBlock2D"]` in the model config and then having to search very deep in `UNet2DCondition` to find them. I would agree here and the unet_blocks design is not the best for readability, but it's an abstraction we "had to chose" in order to not make the whole code base explode and to be able to re-use code better. We are not keen on having more parts of the code be like this which is why I'm not in favor of the "_type" logic for the attention layers/classes.

Hope this makes sense :-) 


 it's arguable not easy to understand what's going on when reading all those

Reply:
Also cc @patil-suraj @williamberman @pcuenca curios to hear your thoughts on this

Reply:
yep!  It totally makes sense:)  will replace it with `add_self_attention`